### PR TITLE
Static method to convert a Map to a valid JSON string.

### PR DESCRIPTION
**Pull request by jsfan3**
I noted that the CodenameOne API is missing of a method to convert a Map<String,Object> to a JSON string, that's why I created a static method to convert the given Map<String,Object> to a valid JSON representation. The values allowed types in the Map<String,Object> are: Number, String, Boolean, List, Map or null. Limited whitespace is inserted be make the resulting JSON string more readable. Simple and complex examples of usage are in the javadoc. I tested the generated JSON strings in a JSON validator. I hope that my proposed method will be helpful.

### CODE REVIEW

File: CodenameOne/src/com/codename1/io/JSONParser.java [hunk 1]
```diff
@@ -632,4 +634,270 @@ public void keyValue(String key, String value) {
...
+     * }
+     * }</pre>
+     *
+     * @param map The map to be converted to a JSON string
+     * @return The JSON string
+     * @throws Exception A generic exception thrown when converting the map to a
+     * JSON string fails.
+     * @author Francesco Galgani
+     */
+    public static String mapToJson(Map<String, Object> map) throws Exception {
```
Comment (Maaartinus) [inline CodenameOne/src/com/codename1/io/JSONParser.java]: - The argument should be `Map<String, ?>`.
- Throwing a checked exception is a big pain for such utils.

File: CodenameOne/src/com/codename1/io/JSONParser.java [hunk 2]
```diff
@@ -632,4 +634,270 @@ public void keyValue(String key, String value) {
...
+        int remainingItems = map.size();
+        for (Map.Entry<String, Object> entry : map.entrySet()) {
+            json.append(indent).append("\"").append(entry.getKey()).append("\" : ");
+            if (entry.getValue() instanceof String) {
+                String value = (String) entry.getValue();
+                json.append("\"").append(value).append("\"");
+            } else if (entry.getValue() instanceof Number) {
+                String value;
+                if (entry.getValue() instanceof Byte) {
+                    value = ((Byte) entry.getValue()).toString();
```
Comment (Maaartinus) [inline CodenameOne/src/com/codename1/io/JSONParser.java]: This cast and all following are needless. You're just testing if it's a `Number` of a kind you accept. Not sure, if it's worth it. What about `Short` or `BigInteger`?

File: CodenameOne/src/com/codename1/io/JSONParser.java [hunk 3]
```diff
@@ -632,4 +634,270 @@ public void keyValue(String key, String value) {
...
+                } else if (entry.getValue() instanceof Double) {
+                    value = ((Double) entry.getValue()).toString();
+                } else if (entry.getValue() instanceof Float) {
+                    value = ((Float) entry.getValue()).toString();
+                } else if (entry.getValue() instanceof Integer) {
+                    value = ((Integer) entry.getValue()).toString();
+                } else if (entry.getValue() instanceof Long) {
+                    value = ((Long) entry.getValue()).toString();
```
Comment (Maaartinus) [inline CodenameOne/src/com/codename1/io/JSONParser.java]: As JS has nothing but `Double` and therefore `Long` can't be represented exactly as number, it's common to offer an option serializing `Long` to `String`.

File: CodenameOne/src/com/codename1/io/JSONParser.java [hunk 4]
```diff
@@ -632,4 +634,270 @@ public void keyValue(String key, String value) {
...
+                } else {
+                    throw new Exception("Cast error in converting map to json");
```
Comment (Maaartinus) [inline CodenameOne/src/com/codename1/io/JSONParser.java]: Such a message isn't helpful. You need the key and the value.

File: CodenameOne/src/com/codename1/io/JSONParser.java [hunk 5]
```diff
@@ -632,4 +634,270 @@ public void keyValue(String key, String value) {
...
+            } else if (entry.getValue() instanceof Boolean) {
+                String value = ((Boolean) entry.getValue()).toString();
+                json.append(value);
+            } else if (entry.getValue() instanceof List) {
+                List myList = (List) (entry.getValue());
+                StringBuilder value = new StringBuilder();
+                value.append("[");
+                if (!myList.isEmpty() && (myList.get(0) instanceof String)) {
+                    for (int i = 0; i < myList.size(); i++) {
+                        String item = (String) (myList.get(i));
```
Comment (Maaartinus) [inline CodenameOne/src/com/codename1/io/JSONParser.java]: - This fails badly for heterogeneous lists.
- It's needlessly repetitive and not very general, what about list of lists? It needs recursion.

File: CodenameOne/src/com/codename1/io/JSONParser.java [hunk 6]
```diff
@@ -632,4 +634,270 @@ public void keyValue(String key, String value) {
...
+                        Boolean item = (Boolean) (myList.get(i));
+                        value.append((Boolean) item);
+                        if (i < myList.size() - 1) {
+                            value.append(",");
+                        }
+                    }
+                } else if (!myList.isEmpty() && myList.get(0) instanceof Map) {
+                    for (int i = 0; i < myList.size(); i++) {
+                        String item = mapToJson((Map) (myList.get(i)));
+                        List<String> lines = StringUtil.tokenize(item, '\n');
```
Comment (Maaartinus) [inline CodenameOne/src/com/codename1/io/JSONParser.java]: Inefficient. Use a helper method like `appendTo(StringBuilder result, Map<String, ?>, int depth)`.

And another one for lists.

File: CodenameOne/src/com/codename1/io/JSONParser.java [hunk 7]
```diff
@@ -632,4 +634,270 @@ public void keyValue(String key, String value) {
...
+                        }
+                        if (i < myList.size() - 1) {
+                            value.append(",");
+                        }
+                    }
+                }
+                value.append("]");
+                json.append(value.toString());
+            } else if (entry.getValue() instanceof Map) {
+                String value = mapToJson((Map) entry.getValue());
```
Comment (Maaartinus) [inline CodenameOne/src/com/codename1/io/JSONParser.java]: What if the keys aren't strings?

**Comment by codenameone** [comment]
You can replace all of that code with:

````java
return Result.fromContent(map).toString();
````

It's not well documented but we use it here and there.

**Comment by jsfan3** [comment]
Thank you "Maaartinus" for the very precise comments.

Thank you "codenameone" (are you Shai?) for solving the problem with one line: I tested your solution against the same Maps that I used with my code and the obtained results are equivalent.
In this new commit, I removed all my code, I inserted your one-line solution, I updated the javadoc with the example outputs resulting from Result.fromContent(map).toString();

Maybe adding this method and its javadoc can be helpfull, because otherwise it's not so intuitive how to produce and convert a Map to a Json string.

**Comment by codenameone** [comment]
Thanks.
Yes I agree this method signature would help document that more effectively.

**Pull request merged**